library/kani_macros/src/lib.rs

```diff
@@ -82,30 +59,44 @@ pub fn async_proof(attr: TokenStream, item: TokenStream) -> TokenStream {
 ///   kani::block_on(harness())
 /// }
 /// ```
-pub fn async_proof(attr: TokenStream, item: TokenStream) -> TokenStream {
-    assert!(attr.is_empty(), "#[kani::async_proof] does not take any arguments for now");
+/// which is then translated again as a asynchronous function.
+#[cfg(kani)]
+#[proc_macro_attribute]
+pub fn proof(attr: TokenStream, item: TokenStream) -> TokenStream {
     let fn_item = parse_macro_input!(item as ItemFn);
     let attrs = fn_item.attrs;
     let vis = fn_item.vis;
     let sig = fn_item.sig;
-    assert!(sig.asyncness.is_some(), "#[kani::async_proof] can only be applied to async functions");
-    assert!(
-        sig.inputs.is_empty(),
-        "#[kani::async_proof] can only be applied to functions without inputs"
-    );
-    let mut modified_sig = sig.clone();
-    modified_sig.asyncness = None;
     let body = fn_item.block;
-    let fn_name = &sig.ident;
-    quote!(
-        #[kani::proof]
-        #(#attrs)*
-        #vis #modified_sig {
-            #sig #body
-            kani::block_on(#fn_name())
-        }
-    )
-    .into()
+
+    assert!(attr.is_empty(), "#[kani::proof] does not take any arguments for now");
+
+    if sig.asyncness.is_none() {
+        quote!(
+            #[kanitool::proof]
+            #[no_mangle]
+            #(#attrs)*
+            #vis #sig #body
+        )
+        .into()
+    } else {
+        assert!(
+            sig.inputs.is_empty(),
+            "#[kani::proof] cannot be applied to async functions that take inputs for now"
+        );
+        let mut modified_sig = sig.clone();
+        modified_sig.asyncness = None;
+        let fn_name = &sig.ident;
+        quote!(
+            #[kani::proof]
```

Comment:
Why not invoke `kanitool::proof` directly? You've already invoked the proc macro and putting it here will cause another level of macro expansion to occur.

Reply:
Because if we change the translation of #[kani::proof] for sync, we may forget to update async, if the code is in different places. This approach seems cleaner to me because it separates two steps: getting rid of async, and then translating to Kani internals. What's the disadvantage of doing it this way? Will performance suffer a lot from this?

Reply:
Well it doesn't have to be defined in multiple places. You can call a subfunction inside of the proc macro.

But, yes, it makes compilation slower since it'll have to invoke this macro multiple times. This includes serializing and deserializing the token tree multiple times.

Reply:
I don't think this matters compared to how long CBMC takes to prove a harness, but I've changed the code to avoid it.